### PR TITLE
Adds functionality for joining a group

### PR DESCRIPTION
**Pull request by vlop11**
Since the actual method for a user joining a group is already written, I will mostly be focusing on the front-end:

- Make the JOIN button for each group call the PUT method that's already written
- Once the button is pressed, render a new button that says JOINED and is not clickable
- Possibly have every already joined group have the above JOINED button

**TODO:**
- Remove all hardcoded strings (example user emails and calendar IDs)

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 1]
```diff
@@ -57,18 +60,20 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
...
             ImmutableMap<String, String> groupMap = ImmutableMap.of(
+                GROUP_ID_PROPERTY, group.getKey().getName(),
                 GROUP_NAME_PROPERTY, (String) group.getProperty(GROUP_NAME_PROPERTY),
                 GROUP_IMAGE_PROPERTY, (String) group.getProperty(GROUP_IMAGE_PROPERTY),
                 GROUP_DESCRIPTION_PROPERTY, (String) group.getProperty(GROUP_DESCRIPTION_PROPERTY));
             groupsListBuilder.add(groupMap);
         }
         
         ImmutableList<ImmutableMap<String, String>> groupsList = groupsListBuilder.build();
+        ImmutableList<Long> userGroups= ImmutableList.copyOf(SoyRendererUtils.getGroupIdList(request));
```
Comment (sporkert) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java]: nit: add a space before the `=`

**Review by sporkert — CHANGES_REQUESTED** [review]

File: src/main/java/com/google/sps/servlets/GroupsServlet.java [hunk 2]
```diff
@@ -139,22 +146,24 @@ private void addUserToGroup(String groupId, String email) throws IOException {
...
 
-        user.setProperty(USER_EMAIL_PROPERTY, email);
+            user.setProperty(USER_EMAIL_PROPERTY, email);
 
-        HashSet<Long> groups = new HashSet<Long>();
-        groups.add(groupIdLong);
-        user.setProperty(GROUPS_KEY, new HashSet<Long>());
+            List<Long> groups = new ArrayList<Long>();
+            groups.add(groupIdLong);
+            user.setProperty(GROUPS_KEY, groups);
```
Comment (courtneywong) [inline src/main/java/com/google/sps/servlets/GroupsServlet.java]: ImmutableList.Builder?

File: src/main/java/com/google/sps/utils/SoyRendererUtils.java [hunk 3]
```diff
@@ -22,4 +34,35 @@ public static String getOutputString(String templateFile, String templateNamespa
 
         return tofu.newRenderer(templateNamespace).setData(templateData).render();
     }
+
+    public static List<Long> getGroupIdList(HttpServletRequest request) {
+        // String userEmail = request.getUserPrincipal().getName();
+        // This is just a hardcoded text email //
+        String userEmail = "example@test.com";
+        // Stores the IDs for all the groups the user ahs joined
+        List<Long> userGroups = new ArrayList<Long>();
```
Comment (courtneywong) [inline src/main/java/com/google/sps/utils/SoyRendererUtils.java]: ImmutableList.Builder?

File: src/main/java/com/google/sps/utils/SoyRendererUtils.java [hunk 4]
```diff
@@ -22,4 +34,35 @@ public static String getOutputString(String templateFile, String templateNamespa
...
+
+        if (user != null) {
+            userGroups = (List<Long>) user.getProperty(GROUPS_KEY);
+
+        } else {
+            // if the user does not exist, create one and add to datastore
+            user = new Entity(USER_KIND, userEmail);
+
+            user.setProperty(USER_EMAIL_PROPERTY, userEmail);
+            // Hardcoded group added to user's list //
```
Comment (courtneywong) [inline src/main/java/com/google/sps/utils/SoyRendererUtils.java]: You don't need the // at the end.

File: src/main/java/com/google/sps/utils/SoyRendererUtils.java [hunk 5]
```diff
@@ -22,4 +34,35 @@ public static String getOutputString(String templateFile, String templateNamespa
...
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query userQuery = new Query(USER_KIND).addFilter(Entity.KEY_RESERVED_PROPERTY, FilterOperator.EQUAL, 
+            KeyFactory.createKey(USER_KIND, userEmail));
+        PreparedQuery userPreparedQuery = datastore.prepare(userQuery);
+        Entity user = userPreparedQuery.asSingleEntity();
+
+        if (user != null) {
+            userGroups = (List<Long>) user.getProperty(GROUPS_KEY);
+
```
Comment (courtneywong) [inline src/main/java/com/google/sps/utils/SoyRendererUtils.java]: remove this newline

File: src/main/webapp/groups.js [hunk 6]
```diff
@@ -10,3 +10,14 @@ function init() {
 function toggleHidden() {
     goog.dom.classlist.toggle(newGroupContainer, "hidden");
 }
+
+function joinGroup(groupId) {
+    fetch("/groups",  {
+        method: "PUT",
+        body: groupId
```
Comment (courtneywong) [inline src/main/webapp/groups.js]: Add a trailing comma here.

File: src/main/webapp/groups.js [hunk 7]
```diff
@@ -10,3 +10,14 @@ function init() {
...
+    });
+
+    const joinButton = goog.dom.getElement("join-btn-" + groupId);
+    goog.dom.setTextContent(joinButton, "Joined");
+    goog.dom.setProperties(joinButton, {"disabled": "true"});
```
Comment (courtneywong) [inline src/main/webapp/groups.js]: Can you use setAttribute here? `joinButton.setAttribute('setAttribute', true)`

File: src/main/webapp/groups.js [hunk 8]
```diff
@@ -10,3 +10,14 @@ function init() {
 function toggleHidden() {
     goog.dom.classlist.toggle(newGroupContainer, "hidden");
 }
+
```
Comment (courtneywong) [inline src/main/webapp/groups.js:13]: Sorry I didn't see this before, but there shouldn't be any double quotation marks in JS. Prefer to use single quotes. Can you fix it throughout teh file? ie "/groups" should be '/groups'.

**Review by courtneywong — CHANGES_REQUESTED** [review]

**Review by ahlschlager — APPROVED** [review]

File: src/main/webapp/groups.js [hunk 9]
```diff
@@ -21,3 +21,23 @@ function joinGroup(groupId) {
...
+    const newGroupId;
+
+    fetch('/groups?name=' + name + '&image=' + image + '&description=' + description, {
+        method: 'POST'
+    }).then(response => response.text()).then(text => {
+            newGroupId = text;
+        }
+    );
+
+    fetch('/calendar?groupId=' + groupId, {
```
Comment (dylancastillo) [inline src/main/webapp/groups.js]: shouldn't this be newGroupId?

**Review by dylancastillo — APPROVED** [review]

**Review by courtneywong — APPROVED** [review]

**Review by sporkert — APPROVED** [review]

**Pull request merged**